src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -2836,8 +2834,31 @@ module Crystal
       false
     end
 
+    def visit(node : TypeRestrict)
+      value = node.value
+      restriction = node.restriction
+
+      value.accept self
+
+      unless context = match_context
```

Comment:
If the default value is a method call. But don't worry, this case won't be possible in the future, and right now it's actually very to trigger that. I think your implementation is more than good, so let's leave it like that :-)